Follow up #33756

bogdanvlviv opened the pull request:
-    Clarify docs of `config.filter_parameters` and `#filter_attributes`
    Add mention that `config.filter_parameters` also filters out sensitive
    values of database columns when call `#inspect` since #33756.

-  DRY `activerecord/lib/active_record/core.rb` and fix tests
   - Move
      ```
      filter_attributes = self.filter_attributes.map(&:to_s).to_set

      filter_attributes.include?(attribute_name) && !read_attribute(attribute_name).nil?
      ```
      to private method.
   - Fix tests in `activerecord/test/cases/filter_attributes_test.rb`
      - Ensure that `teardown` sets `ActiveRecord::Base.filter_attributes` to
        previous state.

    - Ensure that `Admin::Account.filter_attributes` is set to previous
        state in the "filter_attributes could be overwritten by models" test.


-  Build string set when `filter_attributes` is assigned
    It would allow `filter_attributes` to be reused across multiple
calls to `#inspect` or `#pretty_print`.
   - Add `require "set"` 
   - Remove `filter_attributes` instance reader. I think there is no need
to keep it.

-    Add mention about `ActiveRecord::Base::filter_attributes` to the changelog entry
    Also remove `# :nodoc:` for `ActiveRecord::Core::ClassMethods` in order
    to show non-nodoc methods in that module on the api docs http://edgeapi.rubyonrails.org


Follow up #33756

bogdanvlviv committed:
Clarify docs of `config.filter_parameters` and `#filter_attributes`

Add mention that `config.filter_parameters` also filters out sensitive
values of database columns when call `#inspect` since #33756.

bogdanvlviv committed:
DRY `activerecord/lib/active_record/core.rb` and fix tests

- Move
  ```
  filter_attributes = self.filter_attributes.map(&:to_s).to_set

  filter_attributes.include?(attribute_name) && !read_attribute(attribute_name).nil?
  ```
  to private method.

- Fix tests in `activerecord/test/cases/filter_attributes_test.rb`
  - Ensure that `teardown` sets `ActiveRecord::Base.filter_attributes` to
    previous state.

  - Ensure that `Admin::Account.filter_attributes` is set to previous
    state in the "filter_attributes could be overwritten by models" test.

Follow up #33756

eugeneius commented on activerecord/lib/active_record/core.rb:
```diff
@@ -583,5 +580,11 @@ def thaw
       end
+
+      def filter_attribute?(attribute_name)
+        filter_attributes = self.filter_attributes.map(&:to_s).to_set
```
We're now doing this work for each attribute, instead of reusing it. I don't think making the implementation slower to avoid some duplication is a good tradeoff.

We could build the string set when `filter_attributes` is assigned, which would allow it to be reused across multiple calls to `#inspect` or `#pretty_print`.

bogdanvlviv replied:
Ugh, I didn't notice that it would work slower. Thanks for the review.

bogdanvlviv committed:
Build string set when `filter_attributes` is assigned

It would allow `filter_attributes` to be reused across multiple
calls to `#inspect` or `#pretty_print`.

- Add `require "set"`
- Remove `filter_attributes` instance reader. I think there is no need
to keep it.

bogdanvlviv commented:
> https://github.com/rails/rails/pull/33832#discussion_r216273533

@eugeneius What do you think about https://github.com/rails/rails/pull/33832/commits/54fc5357c814aeeea56914782d6613e43b1e53fa?

eugeneius commented:
Looks great 👍

eugeneius commented on activerecord/lib/active_record/core.rb:
```diff
@@ -227,6 +226,20 @@ def generated_association_methods
         end
       end
 
+      # Returns columns which shouldn't be exposed while calling #inspect.
```
Should these methods be documented? They're not currently, because RDoc doesn't pick up [metaprogrammed methods](http://docs.seattlerb.org/rdoc/RDoc/Parser/Ruby.html#class-RDoc::Parser::Ruby-label-Metaprogrammed+Methods) inside an `included do` block, but that seems unintentional:
https://edgeapi.rubyonrails.org/classes/ActiveRecord/Core.html

If `filter_attributes` is a documented feature that users should set directly, it should have a changelog entry and be mentioned in the guides; but if it's an implementation detail and users should only ever set `filter_parameters`, then these accessors should be `:nodoc:`.

bogdanvlviv replied:
I think `filter_attributes` should be public API. We have even added tests for that https://github.com/rails/rails/blob/master/activerecord/test/cases/filter_attributes_test.rb#L32. Yes, it makes sense to add changelog entry too.

> RDoc doesn't pick up metaprogrammed methods inside an included do block, but that seems unintentional:
https://edgeapi.rubyonrails.org/classes/ActiveRecord/Core.html

Yes, I'm seeing the same problem. At least we should remove `# :nodoc:` for `ClassMethods` module in order to see the non-nodoc methods in the api docs. Does it make sense?

eugeneius replied:
It looks like the module is `:nodoc:` intentionally: https://github.com/rails/rails/pull/30146

We could remove `:nodoc:` from the individual methods to make it clearer that it's not a mistake, but that seems unrelated to the rest of the changes here.

bogdanvlviv replied:
Since we want to see `filter_attributes` in the api docs, I think I should remove `# :nodoc:` for `ClassMethods`  and add `# :nodoc:` to the methods in that module except `filter_attributes`/`=`.

But I'm not completly sure because maybe there is reason why we that can't remove `nodoc` for the `ClassMethods` since #30146?
/cc @kamipo

eugeneius replied:
Oh, I missed that the new methods here were in the `ClassMethods` module. Your suggestion makes sense to me, but I'd like to hear @kamipo's opinion too.

bogdanvlviv committed:
Add mention about `ActiveRecord::Base::filter_attributes` to the changelog entry

Also remove `# :nodoc:` for `ActiveRecord::Core::ClassMethods` in order
to show non-nodoc methods in that module on the api docs http://edgeapi.rubyonrails.org

ricardotk002 commented on activerecord/CHANGELOG.md:
```diff
@@ -1,8 +1,13 @@
-*   Configuration item `config.filter_parameters` could also filter out sensitive value of database column when call `#inspect`.
+*   Configuration item `config.filter_parameters` could also filter out
+    sensitive values of database columns when call `#inspect`.
+    We aslo added `ActiveRecord::Base::filter_attributes`/`=` in order to
```
A small typo here: "We **also**..."

bogdanvlviv replied:
Thanks! I just fixed it.
